### PR TITLE
Fix FIFO enforcement to maintain a list

### PR DESCRIPTION
**Pull request by poornas**
of oldest objects qualifying for deletion.
Use this list when bucket exceeds configured FIFO
quota to bring it within quota.

## Motivation and Context
Make FIFO quota enforcement more efficient by maintaining a list of objects eligible for deletion for buckets configured with FIFO quota.

## How to test this PR?
Set up a bucket with FIFO quota using `mc admin bucket quota` , and do `mc cp` to fill objects in bucket over configured quota. When crawler runs, oldest objects over quota should be cleaned up.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [ ] Unit tests needed
- [ ] Functional tests needed (If yes, add [mint](https://github.com/minio/mint) PR # here: )

### CODE REVIEW

File: cmd/bucket-quota.go [hunk 1]
```diff
@@ -29,9 +31,42 @@ import (
 	"github.com/minio/minio/pkg/madmin"
 )
 
+const (
+	fifoDeletionFile = "fifodelete.json"
```
Comment (harshavardhana) [inline cmd/bucket-quota.go:37]: no json use `msgpack` :-) @poornas

File: cmd/bucket-quota.go [hunk 2]
```diff
@@ -49,7 +84,47 @@ func (sys *BucketQuotaSys) Get(bucketName string) (*madmin.BucketQuota, error) {
...
+
+// Init initializes the bucket storage cache.
+func (sys *BucketQuotaSys) Init(ctx context.Context, objAPI ObjectLayer) error {
+	if env.Get(envDataUsageCrawlConf, config.EnableOn) == config.EnableOff {
+		return nil
+	}
+	c := sys.cache
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	dui, err := loadDataUsageFromBackend(ctx, objAPI)
```
Comment (harshavardhana) [inline cmd/bucket-quota.go:102]: don't lock loading from disk @poornas

File: cmd/bucket-quota.go [hunk 3]
```diff
@@ -49,7 +84,47 @@ func (sys *BucketQuotaSys) Get(bucketName string) (*madmin.BucketQuota, error) {
...
+	if err != nil {
+		return err
+	}
+	buckets, err := objAPI.ListBuckets(ctx)
```
Comment (harshavardhana) [inline cmd/bucket-quota.go:106]: don't lock to buckets from disk

File: cmd/bucket-quota.go [hunk 4]
```diff
@@ -140,88 +200,59 @@ func startBucketQuotaEnforcement(ctx context.Context, objAPI ObjectLayer) {
...
-			return err
-		}
-
-		// reuse the fileScorer used by disk cache to score entries by
-		// ModTime to find the oldest objects in bucket to delete. In
-		// the context of bucket quota enforcement - number of hits are
-		// irrelevant.
-		scorer, err := newFileScorer(toFree, time.Now().Unix(), 1)
+		// fetch list of oldest objects in bucket eligible for deletion
+		quotaDelFile := pathJoin(bucketMetaPrefix, bucket, fifoDeletionFile)
```
Comment (harshavardhana) [inline cmd/bucket-quota.go:224]: I dont think you can create new files in this manner, what is the purpose? @poornas can you re-implement this to use crawler cache to mark files for deletion?

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#6fd82c88-c9fb-4e06-a8d1-2871714fb89a) |
<a name="6fd82c88-c9fb-4e06-a8d1-2871714fb89a"></a>
### 9751-56b7e2f/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev7.minio.io:31873
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp e689d84a4ec7:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 4 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 46 seconds
(4/15) Running aws-sdk-ruby tests ... done in 10 seconds
(5/15) Running awscli tests ... done in 2 minutes and 39 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 40 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 2 seconds
(9/15) Running minio-go tests ... done in 1 minutes and 47 seconds
(10/15) Running minio-java tests ... FAILED in 5 minutes and 29 seconds
{
  "name": "minio-java",
  "function": "removeIncompleteUpload(String bucketName, String objectName)",
  "duration": 1347,
  "status": "FAIL",
  "error": "java.io.EOFException >>> [okio.RealBufferedSink.write(RealBufferedSink.java:115), io.minio.HttpRequestBody.writeTo(HttpRequestBody.java:78), okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:83), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147), okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121), okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:94), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121), okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147), okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:125), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121), okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:264), okhttp3.RealCall.execute(RealCall.java:93), io.minio.MinioClient.execute(MinioClient.java:1097), io.minio.MinioClient.execute(MinioClient.java:1236), io.minio.MinioClient.executePut(MinioClient.java:1410), io.minio.MinioClient.executePut(MinioClient.java:1424), io.minio.MinioClient.uploadPart(MinioClient.java:5492), io.minio.MinioClient.putObject(MinioClient.java:3740), io.minio.MinioClient.putObject(MinioClient.java:3870), FunctionalTest.removeIncompleteUploads_test(FunctionalTest.java:1878), FunctionalTest.runTests(FunctionalTest.java:4055), FunctionalTest.main(FunctionalTest.java:4192)]"
}
(10/15) Running minio-js tests ... done in 54 seconds
(11/15) Running minio-py tests ... done in 4 minutes and 34 seconds
(12/15) Running s3cmd tests ... done in 29 seconds
(13/15) Running s3select tests ... done in 12 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

File: cmd/bucket-quota.go (hunk 1 above)
Comment (poornas) [reply cmd/bucket-quota.go:37]: just writing a slice of string, hence  do not see the point of utilizing msgpack here.

File: cmd/bucket-quota.go (hunk 4 above)
Comment (poornas) [reply cmd/bucket-quota.go:224]: fifo quota enforcement needs to build a list of files sorted by age when bucket usage hits a certain threshold. This way, when bucket exceeds quota, the enforcement routine will kick in and use this list to clear oldest entries until usage is within quota.AB asked that this list should be in its own directory under the .minio.sys/buckets/bucketname hierarchy.

Comment (harshavardhana) [reply cmd/bucket-quota.go:224]: we should discuss this further, I think adding more and more files like this is going to make backend cryptic. We should re-use what we have perhaps and re-implement FIFO first to use the crawler framework.

File: cmd/bucket-quota.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/bucket-quota.go:37]: msgpack is compressible, efficient for these purposes using JSON doesn't make sense anymore for these purposes.

File: cmd/bucket-quota.go (hunk 4 above)
Comment (harshavardhana) [reply cmd/bucket-quota.go:224]: we should also see if you can re-use the information present in the bloomfilter here, we are like capturing similar information at multiple places it is not going to be easy to debug things in this manner. 

Lets talk to @klauspost and see how we can combine them for similar purposes.

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

File: cmd/bucket-quota.go (hunk 1 above)
Comment (poornas) [reply cmd/bucket-quota.go:37]: changed this.

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#e50cb0f0-a02c-437e-94b7-ca6753f65bbc) |
<a name="e50cb0f0-a02c-437e-94b7-ca6753f65bbc"></a>
### 9751-87ac7de/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:31111
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 4e910f0266bf:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... FAILED in 4 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-uic9urwicm0v62",
    "objectName": "pwd5nz4amim9iukboaraphaycvagzc"
  },
  "duration": 702,
  "error": "RequestError: send request failed\ncaused by: Put http://minio-dev6.minio.io:31111/aws-sdk-go-test-uic9urwicm0v62/pwd5nz4amim9iukboaraphaycvagzc?tagging=: EOF",
  "function": "testObjectTagging",
  "message": "AWS SDK Go PUTObjectTagging expected to success but got RequestError",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(1/15) Running aws-sdk-java tests ... done in 2 seconds
(2/15) Running aws-sdk-php tests ... done in 45 seconds
(3/15) Running aws-sdk-ruby tests ... done in 7 seconds
(4/15) Running awscli tests ... FAILED in 53 seconds
{
  "name": "awscli",
  "duration": 14528,
  "function": "aws --endpoint-url http://minio-dev6.minio.io:31111 s3api copy-object --bucket awscli-mint-test-bucket-25060 --storage-class REDUCED_REDUNDANCY --key datafile-1-kB --copy-source awscli-mint-test-bucket-25060/datafile-1-kB\n",
  "status": "FAIL",
  "error": "An error occurred (SlowDown) when calling the CopyObject operation (reached max retries: 4): Please reduce your request"
}
(4/15) Running healthcheck tests ... done in 0 seconds
(5/15) Running mc tests ... FAILED in 2 minutes and 17 seconds
{
  "name": "mc",
  "duration": "113434",
  "function": "test_mirror_list_objects_storage_class",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-1957 --quiet --no-color mirror --storage-class REDUCED_REDUNDANCY /mint/data myminio/mc-test-bucket-19716  >>> \n`/mint/data/datafile-1-MB` -> `myminio/mc-test-bucket-19716/datafile-1-MB`\n`/mint/data/datafile-0-b` -> `myminio/mc-test-bucket-19716/datafile-0-b`\n`/mint/data/datafile-1-b` -> `myminio/mc-test-bucket-19716/datafile-1-b`\n`/mint/data/datafile-1-kB` -> `myminio/mc-test-bucket-19716/datafile-1-kB`\n`/mint/data/datafile-1.03-MB` -> `myminio/mc-test-bucket-19716/datafile-1.03-MB`\n`/mint/data/datafile-10-kB` -> `myminio/mc-test-bucket-19716/datafile-10-kB`\n`/mint/data/datafile-10-MB` -> `myminio/mc-test-bucket-19716/datafile-10-MB`\n`/mint/data/datafile-11-MB` -> `myminio/mc-test-bucket-19716/datafile-11-MB`\n`/mint/data/datafile-100-kB` -> `myminio/mc-test-bucket-19716/datafile-100-kB`\n`/mint/data/datafile-129-MB` -> `myminio/mc-test-bucket-19716/datafile-129-MB`\n`/mint/data/datafile-5-MB` -> `myminio/mc-test-bucket-19716/datafile-5-MB`\n`/mint/data/datafile-33-kB` -> `myminio/mc-test-bucket-19716/datafile-33-kB`\n`/mint/data/datafile-5243880-b` -> `myminio/mc-test-bucket-19716/datafile-5243880-b`\n`/mint/data/datafile-6-MB` -> `myminio/mc-test-bucket-19716/datafile-6-MB`\n`/mint/data/datafile-65-MB` -> `myminio/mc-test-bucket-19716/datafile-65-MB`\nmc: <ERROR> Failed to copy `/mint/data/datafile-5243880-b`. Please reduce your request\nmc: <ERROR> Failed to copy `/mint/data/datafile-11-MB`. Please reduce your request\nmc: <ERROR> Failed to copy `/mint/data/datafile-1-kB`. Please reduce your request\nmc: <ERROR> Failed to copy `/mint/data/datafile-1-MB`. Please reduce your request\nmc: <ERROR> Failed to copy `/mint/data/datafile-1-b`. Please reduce your request\nmc: <ERROR> Failed to copy `/mint/data/datafile-1.03-MB`. Please reduce your request\nTotal: 0 B, Transferred: 628.64 MiB, Speed: 5.56 MiB/s"
}
(5/15) Running minio-dotnet tests ... done in 1 minutes and 10 seconds
(6/15) Running minio-go tests ... FAILED in 3 minutes and 1 seconds
{
  "args": {
    "bucketName": "minio-go-test-ebbsssmv6mexrrd0",
    "objectPrefix": "",
    "recursive": "true"
  },
  "duration": 88560,
  "error": "Please reduce your request",
  "function": "ListObjects(bucketName, objectPrefix, recursive, doneCh)",
  "message": "PutObject 4 call failed",
  "name": "minio-go: testListObjects",
  "status": "FAIL"
}
(6/15) Running minio-java tests ... FAILED in 5 seconds
{
  "name": "minio-java",
  "function": "putObject(String bucketName, String objectName, InputStream stream, PutObjectOptions options)",
  "args": "X-Amz-Storage-Class: REDUCED_REDUNDANCY",
  "duration": 34,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = SlowDown, message = Please reduce your request, bucketName = minio-java-test-t7orkf, objectName = minio-java-test-3cmi3o3, resource = /minio-java-test-t7orkf/minio-java-test-3cmi3o3, requestId = 1614E34EBCA2980C, hostId = afc5cef3-b21b-46ad-b01c-640905083e33)\nrequest={method=POST, url=http://minio-dev6.minio.io:31111/minio-java-test-t7orkf/minio-java-test-3cmi3o3?uploads=, headers=X-Amz-Storage-Class: REDUCED_REDUNDANCY\nContent-Type: application/octet-stream\nHost: minio-dev6.minio.io:31111\nAccept-Encoding: identity\nUser-Agent: MinIO (amd64; amd64) minio-java/dev\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20200603T010815Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20200603/us-east-1/s3/aws4_request, SignedHeaders=accept-encoding;content-md5;host;x-amz-content-sha256;x-amz-date;x-amz-storage-class, Signature=*REDACTED*\n}\nresponse={code=503, headers=Accept-Ranges: bytes\nContent-Length: 361\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nRetry-After: 120\nServer: MinIO/DEVELOPMENT.GOGET\nVary: Origin\nX-Amz-Request-Id: 1614E34EBCA2980C\nX-Xss-Protection: 1; mode=block\nDate: Wed, 03 Jun 2020 01:08:15 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:1211), io.minio.MinioClient.execute(MinioClient.java:1236), io.minio.MinioClient.executePost(MinioClient.java:1388), io.minio.MinioClient.createMultipartUpload(MinioClient.java:5092), io.minio.MinioClient.putObject(MinioClient.java:3708), io.minio.MinioClient.putObject(MinioClient.java:3870), FunctionalTest.putObject_test9(FunctionalTest.java:697), FunctionalTest.runTests(FunctionalTest.java:4027), FunctionalTest.main(FunctionalTest.java:4192)]"
}
(6/15) Running minio-js tests ... done in 53 seconds
(7/15) Running minio-py tests ... done in 4 minutes and 11 seconds
(8/15) Running s3cmd tests ... done in 25 seconds
(9/15) Running s3select tests ... done in 12 seconds
(10/15) Running security tests ... done in 0 seconds

Executed 10 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#921af7c8-db0b-4cb0-b169-16e1edeb1749) |
<a name="921af7c8-db0b-4cb0-b169-16e1edeb1749"></a>
### 9751-7943d15/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c3.minio.io:30517
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 38ee85569723:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 0 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 41 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... done in 2 minutes and 12 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 45 seconds
(8/15) Running minio-dotnet tests ... done in 31 seconds
(9/15) Running minio-go tests ... done in 58 seconds
(10/15) Running minio-java tests ... done in 51 seconds
(11/15) Running minio-js tests ... done in 54 seconds
(12/15) Running minio-py tests ... done in 2 minutes and 23 seconds
(13/15) Running s3cmd tests ... FAILED in 8 seconds
{
  "name": "s3cmd",
  "duration": "3328",
  "function": "test_get_object_multipart",
  "status": "FAIL",
  "error": "\n!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n    An unexpected error has occurred.\n  Please try reproducing the error using\n  the latest s3cmd code from the git master\n  branch found at:\n    https://github.com/s3tools/s3cmd\n  and have a look at the known issues list:\n    https://github.com/s3tools/s3cmd/wiki/Common-known-issues-and-their-solutions\n  If the error persists, please report the\n  following lines (removing any private\n  info as necessary) to:\n   s3tools-bugs@lists.sourceforge.net\n\n\n!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n\nInvoked as: /usr/local/bin/s3cmd --config /tmp/.s3cmd-14024/s3cfg get s3://s3cmd-test-bucket-9285/s3cmd-test-object-31540 s3cmd-test-object-31540.downloaded\nProblem: <class 'ConnectionResetError: [Errno 104] Connection reset by peer\nS3cmd:   2.1.0\npython:   3.6.9 (default, Apr 18 2020, 01:56:04) \n[GCC 8.4.0]\nenvironment LANG=C.UTF-8\n\nTraceback (most recent call last):\n  File \"/usr/local/bin/s3cmd\", line 3121, in <module>\n    rc = main()\n  File \"/usr/local/bin/s3cmd\", line 3030, in main\n    rc = cmd_func(args)\n  File \"/usr/local/bin/s3cmd\", line 604, in cmd_object_get\n    response = s3.object_get(uri, dst_stream, destination, start_position = start_position, extra_label = seq_label)\n  File \"/usr/local/lib/python3.6/dist-packages/S3/S3.py\", line 726, in object_get\n    response = self.recv_file(request, stream, labels, start_position)\n  File \"/usr/local/lib/python3.6/dist-packages/S3/S3.py\", line 1733, in recv_file\n    data = http_response.read(this_chunk)\n  File \"/usr/lib/python3.6/http/client.py\", line 459, in read\n    n = self.readinto(b)\n  File \"/usr/lib/python3.6/http/client.py\", line 503, in readinto\n    n = self.fp.readinto(b)\n  File \"/usr/lib/python3.6/socket.py\", line 586, in readinto\n    return self._sock.recv_into(b)\nConnectionResetError: [Errno 104] Connection reset by peer\n\n!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n    An unexpected error has occurred.\n  Please try reproducing the error using\n  the latest s3cmd code from the git master\n  branch found at:\n    https://github.com/s3tools/s3cmd\n  and have a look at the known issues list:\n    https://github.com/s3tools/s3cmd/wiki/Common-known-issues-and-their-solutions\n  If the error persists, please report the\n  above lines (removing any private\n  info as necessary) to:\n   s3tools-bugs@lists.sourceforge.net\n!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!"
}
(13/15) Running s3select tests ... done in 3 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

File: cmd/bucket-quota.go (hunk 4 above)
Comment (klauspost) [reply cmd/bucket-quota.go:224]: Why are we doing this?

It is extremely superflous to have all of this when it should just be part of the crawler.

I thought we agreed that this was just a temporary solution that, while fundamentally broken, was just a stopgap until we could do this properly?

I might as well stop waiting for versioning and push on with more crawler changes.

Comment (klauspost) [reply cmd/bucket-quota.go:224]: oh yeah, the bloom filters are linked to the crawler cycles, so we have strict cutoff points where we move to a clean BF. Another crawler cannot hook into them without breaking the synchronization.

Comment (poornas) [reply cmd/bucket-quota.go:224]: @klauspost, if crawler can give back sorted list of oldest files in a bucket up to certain size that will work for FIFO enforcement. Per discussion with @abperiasamy , the other approach discussed was to use a callback from the crawler - when a FIFO bucket approaches 80% of configured quota, it registers to get metadata for the bucket when the crawler runs and FIFO routine then builds its sorted list from this metadata and saves a list of files eligible for deletion in .minio.sys/buckets/bucket hierarchy.This is what has been implemented, however if it is cleaner to handle all this within crawler that would be nicer.

**Pull request closed without merging**